src/Symfony/Component/DependencyInjection/Exception/AutowiringFailedException.php

```diff
@@ -23,7 +23,10 @@ public function __construct(string $serviceId, $message = '', int $code = 0, \Th
     {
         $this->serviceId = $serviceId;
 
-        if ($message instanceof \Closure && \function_exists('xdebug_is_enabled') && xdebug_is_enabled()) {
+        if (
+            $message instanceof \Closure
+            && (\function_exists('xdebug_is_enabled') && xdebug_is_enabled() || \function_exists('xdebug_info'))
+        ) {
             $message = $message();
```

Comment:
@derickr FYI we have this check here because xdebug changes the behavior of PHP in a way that prevents us from doing a lazy evaluation here: native PHP allows one to set the $message property of throwables to a stringable. We use this below.
It would be great if debug could preserve this behavior, if possible.

Reply:
I think i came up with a better solution. Please check the second commit, which replaces all xdebug-related stuff with resolving the messageCallback when the __toString method is called.
This will also eliminate performance drawbacks when xdebug is installed.

Reply:
> I think i came up with a better solution. Please check the second commit, which replaces all xdebug-related stuff with resolving the messageCallback when the __toString method is called.
> This will also eliminate performance drawbacks when xdebug is installed.

I take this back. `getMessage()` does not return the actual message.